src/Aspire.Dashboard/Telemetry/DashboardTelemetryService.cs

```diff
@@ -0,0 +1,390 @@
+            _telemetryEnabled = await telemetrySender.TryStartTelemetrySessionAsync().ConfigureAwait(false);
+            logger.LogDebug("Initialized telemetry service. Telemetry enabled: {TelemetryEnabled}", _telemetryEnabled);
+
+            // Post session property values after initialization, if telemetry has been enabled.
+            if (_telemetryEnabled is true)
+            {
+                foreach (var (key, value) in GetDefaultProperties())
+                {
+                    PostProperty(key, value);
+                }
+            }
+        }
+        finally
+        {
+            _lock.Release();
+        }
+    }
+
+    private bool IsEnabled()
+    {
+        if (_telemetryEnabled is null)
+        {
+            throw new InvalidOperationException("InitializeAsync has not been called yet");
+        }
+        return _telemetryEnabled.Value;
+    }
+
+    /// <summary>
+    /// Begin a long-running user operation. Prefer this over <see cref="PostOperation"/>. If an explicit user task caused this operation to start,
+    /// use <see cref="StartUserTask"/> instead. Duration will be automatically calculated and the end event posted after <see cref="DashboardTelemetryService.EndOperation"/> is called.
+    /// </summary>
+    public OperationContext StartOperation(string eventName, Dictionary<string, AspireTelemetryProperty> startEventProperties, TelemetrySeverity severity = TelemetrySeverity.Normal, bool isOptOutFriendly = false, bool postStartEvent = true, IEnumerable<OperationContextProperty>? correlations = null)
+    {
+        if (!IsEnabled())
+        {
+            return OperationContext.Empty;
+        }
+
+        var context = OperationContext.Create(propertyCount: 2, name: GetCompositeEventName(eventName, TelemetryEndpoints.TelemetryStartOperation));
+        telemetrySender.QueueRequest(context, async (client, propertyGetter) =>
+        {
+            var scopeSettings = new AspireTelemetryScopeSettings(
+                startEventProperties,
+                severity,
+                isOptOutFriendly,
+                correlations?.Select(propertyGetter).Cast<TelemetryEventCorrelation>().ToArray(),
+                postStartEvent);
+
+            var response = await PostRequestAsync<StartOperationRequest, StartOperationResponse>(client, TelemetryEndpoints.TelemetryStartOperation, new StartOperationRequest(eventName, scopeSettings)).ConfigureAwait(false);
+            context.Properties[0].SetValue(response.OperationId);
+            context.Properties[1].SetValue(response.Correlation);
+        });
+
+        return context;
+    }
+
+    /// <summary>
+    /// Ends a long-running operation. This will post the end event and calculate the duration.
+    /// </summary>
+    public void EndOperation(OperationContextProperty? operationId, TelemetryResult result, string? errorMessage = null)
+    {
+        if (!IsEnabled() || operationId is null)
```

Comment:
Seems odd to have the method noop if there isn't an operationId. Why not have the caller check if there is an operationId and make that decision?

Some with EndUserTask.